### PR TITLE
[Draw/walk area task] Prevent user from drawing self-intersecting polygons

### LINKED ISSUE #3234: Mis-shapen polygon can be drawn through "Complete"
Related to https://github.com/google/ground-android/issues/2604

Adding a point during polygon drawing that create a mis-shapen polygon correctly gives an error. However, if adding it as the last point ("Complete") the mis-shapen polygon goes through.

https://github.com/user-attachments/assets/687965e5-8f9a-4f54-be50-89f071c38b5e

Discovered by @lecrabe, who asked why we even prohibit such polygons?

@gino-m @anandwana001 @shobhitagarwal1612 @kenstershiro

### PR DESCRIPTION
**Pull request by anandwana001**
Fixes #3234 

This PR introduces a full system to detect and handle polygon self-intersections:

Implemented isSelfIntersecting(vertices) to check for crossing edges, with support for both open polylines and closed polygons.

Integrated validation into the workflow:
Add Point: checks the newly added edge, reverts the vertex, and shows a dialog if self-intersection is detected.
Complete Polygon: validates the would-be closed ring (vertices + first), blocking completion and showing a dialog if invalid.
Ensured the intersection flag resets when geometry is valid, preventing stale “intersecting” states.
Updated onFeatureUpdated to disable the Complete button when a polygon is invalid.

Added unit tests covering:
Open and closed polygons
Adjacent edges sharing vertices
True self-intersections (e.g., “X” shapes)
Edge cases like empty list, single point, or triangles

This provides reliable guardrails to prevent creating invalid polygons and ensures user feedback when intersections occur.

https://github.com/user-attachments/assets/25c32662-d3df-48ec-bcde-4b366db1e214

@gino-m @jo-spek  PTAL?

### CODE REVIEW

File: app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskViewModel.kt [hunk 1]
```diff
@@ -95,8 +95,12 @@ internal constructor(
   private val _showSelfIntersectionDialog = MutableSharedFlow<Unit>()
   val showSelfIntersectionDialog = _showSelfIntersectionDialog.asSharedFlow()
 
+  private var hasSelfIntersection: Boolean = false
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskViewModel.kt]: We can declare the setter as private and getter as public. Then we won't need a separate `hasSelfIntersection()` method

File: app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskViewModel.kt [hunk 2]
```diff
@@ -245,11 +241,38 @@ internal constructor(
...
-  fun checkVertexIntersection() {
-    if (isSelfIntersecting(vertices)) {
+  fun checkVertexIntersection(): Boolean {
+    val hit = isSelfIntersecting(vertices)
+    hasSelfIntersection = hit
+    if (hit) {
       vertices = vertices.dropLast(1)
       onSelfIntersectionDetected()
+    } else {
+      hasSelfIntersection = false
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskViewModel.kt]: Isn't this condition implicit? Any reason for explicitly assigning it to false?

File: app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskViewModel.kt [hunk 3]
```diff
@@ -245,11 +241,38 @@ internal constructor(
...
+    }
+    return hit
+  }
+
+  fun validateBeforeComplete(): Boolean {
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskViewModel.kt]: Please consider using a better name for the function. Maybe something which clarifies the purpose of the method.

File: app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskViewModel.kt [hunk 4]
```diff
@@ -245,11 +241,38 @@ internal constructor(
...
+        vertices
+      }
+
+    val hit = isSelfIntersecting(ring)
+    hasSelfIntersection = hit
+    if (hit) {
+      onSelfIntersectionDetected()
+      return false
     }
+    hasSelfIntersection = false
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskViewModel.kt]: Isn't this statement also implicit?

File: app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskViewModel.kt [hunk 5]
```diff
@@ -245,11 +241,38 @@ internal constructor(
...
+
+  fun validateBeforeComplete(): Boolean {
+    val ring =
+      if (vertices.size >= 3 && vertices.first() != vertices.last()) {
+        vertices + vertices.first()
+      } else {
+        vertices
+      }
+
+    val hit = isSelfIntersecting(ring)
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskViewModel.kt]: Can we reuse `hasSelfIntersection` instead of defining another local variable?

File: app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskFragment.kt [hunk 6]
```diff
@@ -92,11 +92,16 @@ class DrawAreaTaskFragment @Inject constructor() : AbstractTaskFragment<DrawArea
...
-        viewModel.triggerVibration()
+        val intersected = viewModel.checkVertexIntersection()
+        if (!intersected) viewModel.triggerVibration()
       }
     completeButton =
-      addButton(ButtonAction.COMPLETE).setOnClickListener { viewModel.completePolygon() }
+      addButton(ButtonAction.COMPLETE).setOnClickListener {
+        if (viewModel.validateBeforeComplete()) {
+          viewModel.completePolygon()
+          viewModel.clearIntersectionFlag()
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskFragment.kt]: Is it even possible to have this flag true when the validation is passing? I feel this line is adding confusion.

File: app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt [hunk 7]
```diff
@@ -85,19 +85,54 @@ private fun onSegment(a: Coordinates, b: Coordinates, c: Coordinates) =
   c.lat in minOf(a.lat, b.lat)..maxOf(a.lat, b.lat) &&
     c.lng in minOf(a.lng, b.lng)..maxOf(a.lng, b.lng)
 
+private data class Segment(val a: Coordinates, val b: Coordinates)
+
+private fun isClosed(vertices: List<Coordinates>): Boolean =
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt]: We already have this method defined inside `Geometry` class. Can we reuse it instead?

File: app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt [hunk 8]
```diff
@@ -85,19 +85,54 @@ private fun onSegment(a: Coordinates, b: Coordinates, c: Coordinates) =
...
+  vertices.size >= 2 && vertices.first() == vertices.last()
+
+private fun shareEndpoint(s1: Segment, s2: Segment): Boolean =
+  s1.a == s2.a || s1.a == s2.b || s1.b == s2.a || s1.b == s2.b
+
+private fun buildSegments(vertices: List<Coordinates>): List<Segment> {
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt]: Please add ktdoc

File: app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt [hunk 9]
```diff
@@ -85,19 +85,54 @@ private fun onSegment(a: Coordinates, b: Coordinates, c: Coordinates) =
...
+  val n = vertices.size
+  if (n < 2) return emptyList()
+  val closed = isClosed(vertices)
+
+  val segs = ArrayList<Segment>(if (closed) n else n - 1) // <- removed unnecessary parentheses
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt]: Please cleanup unused comments.

File: app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt [hunk 10]
```diff
@@ -85,19 +85,54 @@ private fun onSegment(a: Coordinates, b: Coordinates, c: Coordinates) =
   c.lat in minOf(a.lat, b.lat)..maxOf(a.lat, b.lat) &&
     c.lng in minOf(a.lng, b.lng)..maxOf(a.lng, b.lng)
 
+private data class Segment(val a: Coordinates, val b: Coordinates)
+
+private fun isClosed(vertices: List<Coordinates>): Boolean =
+  vertices.size >= 2 && vertices.first() == vertices.last()
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt]: Should this condition be `> 3`?

File: app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt [hunk 11]
```diff
@@ -85,19 +85,54 @@ private fun onSegment(a: Coordinates, b: Coordinates, c: Coordinates) =
...
+private fun buildSegments(vertices: List<Coordinates>): List<Segment> {
+  val n = vertices.size
+  if (n < 2) return emptyList()
+  val closed = isClosed(vertices)
+
+  val segs = ArrayList<Segment>(if (closed) n else n - 1) // <- removed unnecessary parentheses
+  for (i in 0 until n - 1) {
+    segs += Segment(vertices[i], vertices[i + 1])
+  }
+  if (closed) segs += Segment(vertices[n - 1], vertices[0])
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt]: Can you add some comments to help add context why this is needed?

File: app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt [hunk 12]
```diff
@@ -85,19 +85,54 @@ private fun onSegment(a: Coordinates, b: Coordinates, c: Coordinates) =
...
+  return segs
+}
+
+private fun areAdjacent(i: Int, j: Int, size: Int, closed: Boolean): Boolean {
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt]: Please add ktdoc and use better parameter names. `i` and `j` should not be used outside of a loop.

File: app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt [hunk 13]
```diff
@@ -85,19 +85,54 @@ private fun onSegment(a: Coordinates, b: Coordinates, c: Coordinates) =
   c.lat in minOf(a.lat, b.lat)..maxOf(a.lat, b.lat) &&
     c.lng in minOf(a.lng, b.lng)..maxOf(a.lng, b.lng)
 
+private data class Segment(val a: Coordinates, val b: Coordinates)
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt]: Please add some ktdoc. What do we mean by a segment?

File: app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt (hunk 7 above)
Comment (anandwana001) [reply app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt]: That is more specific to LineString object; here we only have a list of coordinates.

File: app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskViewModel.kt [hunk 14]
```diff
@@ -245,11 +240,34 @@ internal constructor(
     }
   }
 
-  fun checkVertexIntersection() {
-    if (isSelfIntersecting(vertices)) {
+  fun checkVertexIntersection(): Boolean {
+    val hit = isSelfIntersecting(vertices)
+    hasSelfIntersection = hit
+    if (hit) {
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskViewModel.kt]: Let's reuse the `hasSelfIntersection` variable here as well instead of using a local variable.

File: app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskViewModel.kt [hunk 15]
```diff
@@ -245,11 +240,34 @@ internal constructor(
...
+    return hit
+  }
+
+  fun validatePolygonCompletion(): Boolean {
+    val ring =
+      if (vertices.size >= 3 && vertices.first() != vertices.last()) {
+        vertices + vertices.first()
+      } else {
+        vertices
+      }
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskViewModel.kt]: We should return early if it's not a polygon.

File: app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskViewModel.kt [hunk 16]
```diff
@@ -245,11 +240,34 @@ internal constructor(
...
+    if (hasSelfIntersection) {
+      onSelfIntersectionDetected()
+      return false
+    }
+    return true
+  }
+
+  fun clearIntersectionFlag() {
+    hasSelfIntersection = false
   }
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskViewModel.kt]: Please remove if unused.

File: app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt (hunk 7 above)
Comment (shobhitagarwal1612) [reply app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt]: The logic below is incorrect. A polygon must have 4 vertices. I'd still suggest somehow reusing the logic instead of duplicating the definition at multiple places.

Comment (anandwana001) [reply app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt]: done

File: app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt [hunk 17]
```diff
@@ -81,23 +109,116 @@ private fun orientation(a: Coordinates, b: Coordinates, c: Coordinates): Int {
...
+  val edges = ArrayList<PolygonEdge>(if (closed) n else n - 1)
+
+  // Create edges between each consecutive pair of vertices.
+  for (i in 0 until n - 1) {
+    edges += PolygonEdge(vertices[i], vertices[i + 1])
+  }
+
+  // If the polygon is closed, add an edge from the last vertex back to the first
+  // to ensure the polygon forms a complete loop.
+  if (closed) edges += PolygonEdge(vertices[n - 1], vertices[0])
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt]: If the polygon is closed, why do we need a edge whose vertices are same? Won't that be of zero length?

File: app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt [hunk 18]
```diff
@@ -81,23 +109,116 @@ private fun orientation(a: Coordinates, b: Coordinates, c: Coordinates): Int {
...
+ * @param edgeCount Total number of edges.
+ * @param isClosedPolygon Whether the polygon is closed (first vertex equals last vertex).
+ */
+private fun areAdjacent(
+  firstEdgeIndex: Int,
+  secondEdgeIndex: Int,
+  edgeCount: Int,
+  isClosedPolygon: Boolean,
+): Boolean {
+  if (firstEdgeIndex == secondEdgeIndex) return true
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt]: Does this condition represent a scenario where the edges being compared are same? If yes, we should prevent it from occurring in the first place itself rather than having to handle it separately.

File: app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt [hunk 19]
```diff
@@ -81,23 +109,116 @@ private fun orientation(a: Coordinates, b: Coordinates, c: Coordinates): Int {
...
+  if (abs(firstEdgeIndex - secondEdgeIndex) == 1) return true
+  if (isClosedPolygon) {
+    val last = edgeCount - 1
+    if (
+      (firstEdgeIndex == 0 && secondEdgeIndex == last) ||
+        (secondEdgeIndex == 0 && firstEdgeIndex == last)
+    ) {
+      return true
+    }
+  }
```
Comment (shobhitagarwal1612) [inline app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt]: We can simplify these method (this and the other ones) by specifying exactly which geometry we are dealing with (i.e. linear ring or polygon). Handling for generic vertices is causing readability issues as you need to check for `completeness` at every step. Either pass a linear ring which ensures that the first and last vertices are same or handle generically for a list of vertices without checking for `completeness`.

**Review by shobhitagarwal1612 — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt (hunk 17 above)
Comment (gino-m) [reply app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt]: +1. If the polygon is closed, I would expect the last edge to end at the first vertex.

File: app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt (hunk 18 above)
Comment (gino-m) [reply app/src/main/java/org/groundplatform/android/ui/util/PolygonUtil.kt]: Aren't polygons represented by the coordinates of the vertices? In that case, the edges are inferred by looking at those coordinates, and there's no way to represent non-adjacent edges.

File: app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskFragment.kt [hunk 20]
```diff
@@ -157,10 +161,12 @@ class DrawAreaTaskFragment @Inject constructor() : AbstractTaskFragment<DrawArea
     val closed = geometry.isClosed()
     val markedComplete = viewModel.isMarkedComplete()
     val tooClose = viewModel.isTooClose.value
+    val selfIntersecting = viewModel.hasSelfIntersection
 
     addPointButton.showIfTrue(!closed)
     addPointButton.enableIfTrue(!closed && !tooClose)
     completeButton.showIfTrue(closed && !markedComplete)
+    completeButton.enableIfTrue(closed && !markedComplete && !selfIntersecting)
```
Comment (gino-m) [inline app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskFragment.kt:169]: How does the user know the "complete button" is disabled because the polygon is self-intersecting? Should we instead be disabling the "add point" button as the user pans the map?

**Comment by gino-m** [comment]
@anandwana001 Note I updated the PR description to reflect the new behavior rather than a description of the code

File: app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskFragment.kt (hunk 20 above)
Comment (shobhitagarwal1612) [reply app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskFragment.kt:169]: +1, we should disable both `add` and `complete` buttons when the polygon is irregular

Comment (shobhitagarwal1612) [reply app/src/main/java/org/groundplatform/android/ui/datacollection/tasks/polygon/DrawAreaTaskFragment.kt:169]: Discussed offline, we are showing a dialog and reverting the last point when adding a point which intersects the polygon

**Review by shobhitagarwal1612 — APPROVED** [review]
